server/runtime/src/main/java/org/infinispan/server/configuration/security/CredentialStoreConfigurationBuilder.java

```diff
@@ -0,0 +1,118 @@
+   private final AttributeSet attributes;
+   private final CredentialStoresConfigurationBuilder credentialStoresConfigurationBuilder;
+   private CredentialStoreSpi credentialStore;
+
+   CredentialStoreConfigurationBuilder(CredentialStoresConfigurationBuilder credentialStoresConfigurationBuilder) {
+      this.attributes = CredentialStoreConfiguration.attributeDefinitionSet();
+      this.credentialStoresConfigurationBuilder = credentialStoresConfigurationBuilder;
+   }
+
+   public CredentialStoreConfigurationBuilder path(String value) {
+      attributes.attribute(PATH).set(value);
+      return this;
+   }
+
+   public CredentialStoreConfigurationBuilder relativeTo(String relativeTo) {
+      attributes.attribute(RELATIVE_TO).set(relativeTo);
+      return this;
+   }
+
+   public CredentialStoreConfigurationBuilder type(String type) {
+      attributes.attribute(TYPE).set(type);
+      return this;
+   }
+
+   public CredentialStoreConfigurationBuilder credential(String credential) {
+      attributes.attribute(CREDENTIAL).set(credential);
+      return this;
+   }
+
+   @Override
+   public void validate() {
+
+   }
+
+   @Override
+   public CredentialStoreConfiguration create() {
+      return new CredentialStoreConfiguration(attributes.protect(), build());
+   }
+
+   @Override
+   public Builder<?> read(CredentialStoreConfiguration template) {
+      this.attributes.read(template.attributes());
+      return this;
+   }
+
+   CredentialStoreSpi build() {
+      if (credentialStore == null) {
+         if (attributes.attribute(PATH).isNull()) {
+            throw new IllegalStateException("file has to be specified");
+         }
+         String path = attributes.attribute(PATH).get();
+         String relativeTo = attributes.attribute(RELATIVE_TO).get();
+         String location = ParseUtils.resolvePath(path, relativeTo);
+         credentialStore = new KeyStoreCredentialStore();
+         final Map<String, String> map = new HashMap<>();
+         map.put("location", location);
+         map.put("create", "false");
+         map.put("keyStoreType", attributes.attribute(TYPE).get());
+         char[] credential = attributes.attribute(CREDENTIAL).get().toCharArray();
+         try {
+            credentialStore.initialize(
+                  map,
+                  new CredentialStore.CredentialSourceProtectionParameter(
+                        IdentityCredentials.NONE.withCredential(new PasswordCredential(ClearPassword.createRaw(ClearPassword.ALGORITHM_CLEAR, credential)))),
+                  null
+            );
+         } catch (CredentialStoreException e) {
+            throw new CacheConfigurationException(e);
+         }
+      }
+      return credentialStore;
+   }
+
+   public <C extends Credential> C getCredential(String alias, Class<C> type) {
+      build();
+      try {
+         if (alias == null) {
+            if (credentialStore.getAliases().size() == 1) {
```

Comment:
would it be better to always fail is `alias==null`? To avoid any chance of returning the wrong credential.

Reply:
You cannot return the wrong credential if there is only one :)